src/eval/metrics.cpp

```diff
@@ -0,0 +1,129 @@
+#include "metrics.h"
+
+namespace Brush {
+namespace Eval {
+
+/* Scoring functions */
+
+/// mean squared error
+float mse(const VectorXf& y, const VectorXf& yhat, VectorXf& loss, 
+            const vector<float>& class_weights)
+{
+    loss = (yhat - y).array().pow(2);
+    return loss.mean(); 
+}
+
+VectorXf log_loss(const VectorXf& y, const VectorXf& predict_proba, 
+                    const vector<float>& class_weights)
+{
+    float eps = pow(10,-10);
+    
+    VectorXf loss;
+    
+    float sum_weights = 0; 
+    loss.resize(y.rows());  
+    for (unsigned i = 0; i < y.rows(); ++i)
+    {
+        if (predict_proba(i) < eps || 1 - predict_proba(i) < eps)
+            // clip probabilities since log loss is undefined for predict_proba=0 or predict_proba=1
+            loss(i) = -(y(i)*log(eps) + (1-y(i))*log(1-eps));
+        else
+            loss(i) = -(y(i)*log(predict_proba(i)) + (1-y(i))*log(1-predict_proba(i)));
+        if (loss(i)<0)
+            std::runtime_error("loss(i)= " + to_string(loss(i)) 
+                    + ". y = " + to_string(y(i)) + ", predict_proba(i) = " 
+                    + to_string(predict_proba(i)));
+
+        if (!class_weights.empty())
+        {
+            loss(i) = loss(i) * class_weights.at(y(i));
+            sum_weights += class_weights.at(y(i));
+        }
+    }
+    
+    if (sum_weights > 0)
+        loss = loss.array() / sum_weights * y.size(); // normalize weight contributions
+    
+    return loss;
+}   
+
+/// log loss
+float mean_log_loss(const VectorXf& y, 
+        const VectorXf& predict_proba, VectorXf& loss,
+        const vector<float>& class_weights)
+{
+        
+    /* std::cout << "loss: " << loss.transpose() << "\n"; */
+    loss = log_loss(y,predict_proba,class_weights);
+    return loss.mean();
+}
+
+
+// multinomial log loss
+VectorXf multi_log_loss(const VectorXf& y, const ArrayXXf& predict_proba, 
```

Comment:
?!

Reply:
This is a placeholder for multi classification, as softmax is not working properly. I will eventually work on this